src/payment/strategies/cardinal/cardinal-three-d-secure-flow.ts

```diff
@@ -0,0 +1,122 @@
+import { some } from 'lodash';
+
+import { CheckoutStore, InternalCheckoutSelectors } from '../../../checkout/index';
+import {
+    MissingDataError,
+    MissingDataErrorType,
+    RequestError
+} from '../../../common/error/errors/index';
+import { default as Payment, CreditCardInstrument } from '../../payment';
+import PaymentActionCreator from '../../payment-action-creator';
+import PaymentMethodActionCreator from '../../payment-method-action-creator';
+
+import { CardinalClient, CardinalOrderData } from './index';
+
+export default class CardinalThreeDSecureFlow {
+    private _clientToken?: string;
+
+    constructor(
+        private _store: CheckoutStore,
+        private _paymentActionCreator: PaymentActionCreator,
+        private _paymentMethodActionCreator: PaymentMethodActionCreator,
+        private _cardinalClient: CardinalClient
+    ) {}
+
+    prepare(methodId: string): Promise<void> {
+        if (this._clientToken) {
+            return Promise.resolve();
+        }
+
+        return this._store.dispatch(this._paymentMethodActionCreator.loadPaymentMethod(methodId))
+            .then(state => {
+                const method = state.paymentMethods.getPaymentMethod(methodId);
+
+                if (!method || !method.config) {
+                    throw new MissingDataError(MissingDataErrorType.MissingPaymentMethod);
+                }
+
+                return this._cardinalClient.initialize(method.config.testMode)
+                    .then(() => {
+                        if (!method.clientToken) {
+                            throw new MissingDataError(MissingDataErrorType.MissingPaymentMethod);
+                        }
+
+                        this._clientToken = method.clientToken;
+
+                        return this._cardinalClient.configure(this._clientToken);
+                    });
+            });
+    }
+
+    start(payment: Payment): Promise<InternalCheckoutSelectors> {
+        if (!payment.paymentData) {
+            throw new MissingDataError(MissingDataErrorType.MissingPayment);
+        }
+
+        const paymentData = payment.paymentData as CreditCardInstrument;
+
+        return this._cardinalClient.runBinProcess(paymentData.ccNumber)
+            .then(() => {
+                if (!this._clientToken) {
+                    throw new MissingDataError(MissingDataErrorType.MissingPaymentMethod);
+                }
+
+                payment = {
+                    ...payment,
+                    paymentData: {
+                        ...paymentData,
+                        threeDSecure: { token: this._clientToken },
```

Comment:
How come we're posting the client token back to BigPay? Doesn't the client token originally come from BigPay?

Reply:
Yes, you're right but we need to send the client token back to `bigpay` due to it contains a unique `ReferenceID` that is not stored in `bigpay` and is formed dynamically each time the client token is generated